src/lib/pipe_lib/replacable_strict_pipe.ml

```diff
@@ -0,0 +1,93 @@
+open Core_kernel
+open Async_kernel
+
+(* NOTE: Data flow:
+      long_live_writer
+   -> long_live_reader
+   -actor-> tmp_swapped_writer
+   -> tmp_swapped_reader (tracked out of this structure)
+*)
+
+type ('data_in_pipe, 'pipe_kind, 'write_return) t =
+  { name : string
+  ; long_live_writer :
+      ('data_in_pipe, 'pipe_kind, 'write_return) Strict_pipe.Writer.t
+  ; long_live_reader : 'data_in_pipe Strict_pipe.Reader.t
+  ; mutable should_terminate : bool
+  ; mutable reader_request :
+      (string * 'data_in_pipe Strict_pipe.Reader.t Ivar.t) option
+  ; mutable tmp_swapped_writer :
+      ( 'data_in_pipe
+      , Strict_pipe.synchronous
+      , unit Deferred.t )
+      Strict_pipe.Writer.t
+      option
+  }
+
+let rec run (t : _ t) =
```

Comment:
Use [`repeat_until_finished`](https://ocaml.janestreet.com/ocaml-core/v0.12/doc/async_kernel/Async_kernel/Deferred/index.html#val-repeat_until_finished) instead of manual loop.